Adv netw addressmap clientside

erosb opened the pull request:
# Fixing client-view of Member#getAddressMap()

## Context

During fixing advanced networking support in MC, we want to use `Member#getAddressMap()` to associate member-to-member addresses to client-to-member addresses. @puzpuzpuz found that on the client side this method returns an incorrect value:

> I can see that the addressMap contains a single entry with EndpointQualifier.MEMBER key and client-to-member address (which seems to be really weird and looks a lot like a bug in Java client).

## Change description

 * The `Member#getAddressMap()` value is populated by `ClientClusterServiceImpl` which didn't pass the `addressMap` from `memberInfo` to `MemberImpl#Builder`, so we fix it
 * `memberInfo.getAddressMap()` was an empty mpa, because the protocol definition didn't cover this attribute, so it is added in the corresponding client-protocol PR, and the generated codecs are included here
 * on the member side, before encoding `MemberInfo` into a `ClientMessage`, we need to copy the `addressMap` from the `Member` into the `MemberInfo` instance, that happens in `ClusterViewListenerService`
 * previously, when called from the client, `Member#getAddress()` returned the client-to-member address. However, `AbstractMember` assumes that if it receives an `addressMap` in its constructor, then it should look up its `address` attribute in the map by key `EndpointQualifier#MEMBER`. This isn't true anymore, since when used on the client side, it should look it up by the EndpointQualifier#CLIENT` key, to retain the previous behavior. Therefore we add a new constructor to `MemberImpl` and `AbstractMember` to make the used lookup key parameterizable, and call it with `CLIENT` key from a new builder method `MemberImpl#Builder#buildForClientView()`.
 * integration test added 
 
 
# Related PRs and issues
 * client protocol change PR: https://github.com/hazelcast/hazelcast-client-protocol/pull/316
 * original problem: https://github.com/hazelcast/management-center/issues/2971
 * PR comment explaining why do we need correct `Member#getAddressMap()` values: https://github.com/hazelcast/management-center/pull/3037#issuecomment-611439434

devOpsHazelcast commented:
[![CLA assistant check](http://ec2-54-211-157-189.compute-1.amazonaws.com:5000/pull/badge/not_signed)](http://ec2-54-211-157-189.compute-1.amazonaws.com:5000/hazelcast/hazelcast?pullRequest=16910) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](http://ec2-54-211-157-189.compute-1.amazonaws.com:5000/hazelcast/hazelcast?pullRequest=16910) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](http://ec2-54-211-157-189.compute-1.amazonaws.com:5000/check/hazelcast/hazelcast?pullRequest=16910) it.</sub>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/193/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/193/Hazelcast-pr-builder-193.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 3 errors reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/cluster/impl/MemberImpl.java:77: Line is longer than 130 characters (found 139). [LineLength]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/client/io/AdvancedNetworkClientIntegrationTest.java:54:15: Unused import - com.hazelcast.instance.ProtocolType.MEMBER. [UnusedImports]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/instance/ProtocolTypeTest.java:1: Line does not match expected header line of '/*'. [Header]
--------------------------
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/194/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/194/Hazelcast-pr-builder-194.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/cluster/impl/MemberImpl.java:78:43: ',' is not followed by whitespace. [WhitespaceAfter]
--------------------------
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/195/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/195/Hazelcast-pr-builder-195.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/201/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/201/Hazelcast-pr-builder-201.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/cluster/impl/MemberImpl.java:208: Don't use trailing comments. [TrailingComment]
--------------------------
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/202/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/202/Hazelcast-pr-builder-202.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   MemberCompatibilityNullTest_2_0.test_ClientAddClusterViewListenerCodec_encodeMembersViewEvent:134->compareClientMessages:7445 Frames have different flags expected:<4096> but was:<2048>
[ERROR]   MemberCompatibilityNullTest_2_1.test_ClientAddClusterViewListenerCodec_encodeMembersViewEvent:134->compareClientMessages:7446 Frames have different flags expected:<4096> but was:<2048>
[ERROR]   MemberCompatibilityTest_2_0.test_ClientAddClusterViewListenerCodec_encodeMembersViewEvent:134->compareClientMessages:7445 Frames have different flags expected:<4096> but was:<2048>
[ERROR]   MemberCompatibilityTest_2_1.test_ClientAddClusterViewListenerCodec_encodeMembersViewEvent:134->compareClientMessages:7446 Frames have different flags expected:<4096> but was:<2048>
[ERROR] Errors: 
[ERROR]   ClientCompatibilityNullTest_2_0.test_ClientAddClusterViewListenerCodec_handleMembersViewEvent:147 ? IllegalArgument
[ERROR]   ClientCompatibilityNullTest_2_1.test_ClientAddClusterViewListenerCodec_handleMembersViewEvent:147 ? IllegalArgument
[ERROR]   ClientCompatibilityTest_2_0.test_ClientAddClusterViewListenerCodec_handleMembersViewEvent:147 ? IllegalArgument
[ERROR]   ClientCompatibilityTest_2_1.test_ClientAddClusterViewListenerCodec_handleMembersViewEvent:147 ? IllegalArgument
[INFO] 
[ERROR] Tests run: 27682, Failures: 4, Errors: 4, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/227/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/227/Hazelcast-pr-builder-227.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/custom/MemberInfoCodec.java:65:9: 'if' is not followed by whitespace. [WhitespaceAround]
--------------------------
</details>
</pre>

sancar commented on hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImpl.java:
```diff
@@ -233,11 +233,12 @@ private void applyInitialState(int version, Collection<MemberInfo> memberInfos)
         LinkedHashMap<UUID, Member> newMembers = new LinkedHashMap<>();
         for (MemberInfo memberInfo : memberInfos) {
-            MemberImpl member = new MemberImpl.Builder(memberInfo.getAddress()).version(memberInfo.getVersion())
+            MemberImpl member = new MemberImpl.Builder(memberInfo.getAddressMap())
```
When this memberInfo is coming from 4.0 cluster, it will be null. The constructor will throw NullPointerException. 
I prefer not to put `buildForClientView` to `MemberImpl.Builder` for the client. The client logic should be handled in client packages. 
Can we check if the addressMap is null, and create the builder with `address` if it is there. 
Otherwise, 



Can we also put a backward compatibility test against 4.1 client against 4.0 cluster to make sure ?

erosb replied:
Hello @sancar , thank you for the review.

> When this memberInfo is coming from 4.0 cluster, it will be null. The constructor will throw NullPointerException.

good catch, fixed

> I prefer not to put buildForClientView to MemberImpl.Builder for the client. The client logic should be handled in client packages.

Agreed, refactored.

> Can we also put a backward compatibility test with 4.1 client against 4.0 cluster to make sure ?

Can you please give me any clue on how to do that?

sancar commented:
@erosb Is this pr deletes the compatibility binaries by mistake ? Can you check it ?

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/229/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/229/Hazelcast-pr-builder-229.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   MemberCompatibilityNullTest_2_1.test_ClientAddClusterViewListenerCodec_encodeMembersViewEvent:134->compareClientMessages:7446 Frames have different flags expected:<4096> but was:<2048>
[ERROR]   MemberCompatibilityTest_2_1.test_ClientAddClusterViewListenerCodec_encodeMembersViewEvent:134->compareClientMessages:7446 Frames have different flags expected:<4096> but was:<2048>
[ERROR] Errors: 
[ERROR]   ClientCompatibilityNullTest_2_1.test_ClientAddClusterViewListenerCodec_handleMembersViewEvent:147 ? IllegalArgument
[ERROR]   ClientCompatibilityTest_2_1.test_ClientAddClusterViewListenerCodec_handleMembersViewEvent:147 ? IllegalArgument
[INFO] 
[ERROR] Tests run: 27721, Failures: 2, Errors: 2, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

mdumandag commented:
I think you need to update the reference objects. For `aListOfMemberInfos`, the binary generator is encoding an `addressMap` but the reference objects does not contain this extra parameter. Also, you need to rebase your branch with recent master. It does not include the MultiMap#putAll changes

erosb commented:
Hello @sancar , thank you for the review.

Regarding your question:

> this pr deletes the compatibility binaries by mistake ? Can you check it ?

While merging `master`, the .binary files conflicted, so during conflict resolution I deleted them and re-generated by running `./generator.py -l java -r ../hazelcast` in the hazelcast-client-protocol repo. Isn't that the right way to do it?
(I can see I also deleted `1.serialization.compatibility.binary`, I will restore that in the next commit).

mdumandag commented:
@erosb I thought this PR was merged but it wasn't. You can rebase your protocol branch with the MultiMap#putAll changes now.
https://github.com/hazelcast/hazelcast-client-protocol/pull/313

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/234/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/234/Hazelcast-pr-builder-234.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   MemberCompatibilityNullTest_2_1.test_ClientAddClusterViewListenerCodec_encodeMembersViewEvent:134->compareClientMessages:7463 Frames have different flags expected:<4096> but was:<2048>
[ERROR]   MemberCompatibilityTest_2_1.test_ClientAddClusterViewListenerCodec_encodeMembersViewEvent:134->compareClientMessages:7463 Frames have different flags expected:<4096> but was:<2048>
[ERROR] Errors: 
[ERROR]   ClientCompatibilityNullTest_2_1.test_ClientAddClusterViewListenerCodec_handleMembersViewEvent:147 ? IllegalArgument
[ERROR]   ClientCompatibilityTest_2_1.test_ClientAddClusterViewListenerCodec_handleMembersViewEvent:147 ? IllegalArgument
[INFO] 
[ERROR] Tests run: 27737, Failures: 2, Errors: 2, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/239/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/239/Hazelcast-pr-builder-239.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   MemberCompatibilityNullTest_2_1.test_ClientAddClusterViewListenerCodec_encodeMembersViewEvent:134->compareClientMessages:7462 Frames have different contents: arrays first differed at element [0]; expected:<108> but was:<0>
[ERROR]   MemberCompatibilityTest_2_1.test_ClientAddClusterViewListenerCodec_encodeMembersViewEvent:134->compareClientMessages:7462 Frames have different contents: arrays first differed at element [0]; expected:<108> but was:<0>
[INFO] 
[ERROR] Tests run: 27737, Failures: 2, Errors: 0, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

mdumandag commented on hazelcast/src/test/java/com/hazelcast/client/protocol/compatibility/ReferenceObjects.java:
```diff
@@ -782,7 +785,9 @@ private static boolean isEqualStackTrace(StackTraceElement stackTraceElement1, S
     private static MemberVersion aMemberVersion = new MemberVersion(aByte, aByte, aByte);
-    public static Collection<MemberInfo> aListOfMemberInfos = Collections.singletonList(new MemberInfo(anAddress, aUUID, aMapOfStringToString, aBoolean, aMemberVersion));
+    public static Collection<MemberInfo> aListOfMemberInfos = Collections.singletonList(new MemberInfo(anAddress, aUUID, aMapOfStringToString, aBoolean, aMemberVersion,
+            ImmutableMap.of(EndpointQualifier.MEMBER, anAddress)));
```
Can you change the ordinal value to 2 at the protocol side to use `ProtocolType.WAN` ? Because the binary encoder is encoding `localhost` as the identifier string for the `EndpointQualifier`. I see that we can specify a custom identifier using `EndpointQualifier#resolve` with the protocol type of `WAN` but it is not an option for other protocol types. 

We can configure binary generator to encode custom values for fix sized types(as you did with ordinal), but there is no way to do so for var sized types in the protocol code generator. I think we should also provide an option for var sized types. I will create an issue for that.

erosb replied:
> Can you change the ordinal value to 2 at the protocol side to use ProtocolType.WAN ?

Great idea, thank you, this helped a lot.

> but there is no way to do so for var sized types

yep, I found the same. I think in `binary/util.py`, `VarSizedParamEncoder#encode_var_sized_frame()` has to be improved to do something similar to `CustomTypeEncoder#create_initial_frame()` (reading from `reference_objects.objects`). To do it, I think `custom_type_name` has to be passed to `#encode_var_sized_frame()` but that method is called from an other place too. I didn't dig deeper today.

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/241/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/241/Hazelcast-pr-builder-241.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/242/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/242/Hazelcast-pr-builder-242.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   MemberImplConstructorTest.testConstructor:42 ? Guardian java.lang.NoSuchMethod...
[INFO] 
[ERROR] Tests run: 27737, Failures: 0, Errors: 1, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

hz-highfive commented:
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/243/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/243/Hazelcast-pr-builder-243.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientQueryCacheRecreationTest.query_cache_recreates_itself_after_server_restart:112->HazelcastTestSupport.assertTrueEventually:1344->HazelcastTestSupport.assertTrueEventually:1242->lambda$query_cache_recreates_itself_after_server_restart$0:114 expected:<200> but was:<191>
[INFO] 
[ERROR] Tests run: 27737, Failures: 1, Errors: 0, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

erosb commented:
verify

erosb commented:
Squashed into a single commit and merged. Thank you @sancar @mdumandag for helping this through.

puzpuzpuz commented:
Does it make sense to backport this fix to `4.0.z`?

erosb commented:
@puzpuzpuz afair we discussed that this can't be backported to 4.0.z because it relies on a client protocol change ( https://github.com/hazelcast/hazelcast-client-protocol/pull/316 )

puzpuzpuz commented:
> @puzpuzpuz afair we discussed that this can't be backported to 4.0.z because it relies on a client protocol change ( https://github.com/hazelcast/hazelcast-client-protocol/pull/316 )

If the change is backward compatible from the protocol perspective, it could be included into a patch release. Sounds like it's a question for the clients team.

@sancar WDYT?

puzpuzpuz commented:
As for the past discussion, we have discussed that MC 4.0 doesn't need this fix. Yet, MC 4.1 has to support both 4.0.z and 4.1.z IMDG versions, so it may make sense to backport the fix.

Of course, that's just a rough idea.

erosb commented:
Well, it adds a [mandatory attribute](https://github.com/hazelcast/hazelcast-client-protocol/pull/316/files#diff-6a97442554fa53caad3f5ee8b5c3d653R785) to `MemberInfo` , so most probably it isn't a backwards-compatible change. I agree with that advanced networking should be fixed in MC 4.1 , so it would be necessary to have addressMap working properly.

puzpuzpuz commented:
> Well, it adds a [mandatory attribute](https://github.com/hazelcast/hazelcast-client-protocol/pull/316/files#diff-6a97442554fa53caad3f5ee8b5c3d653R785) to `MemberInfo` , so most probably it isn't a backwards-compatible change.

That attribute also has `since: 2.1` definition, so the generated codecs [assume](https://github.com/hazelcast/hazelcast/pull/16910/files#diff-85119ef45bcf87c56a3c4e6485d02c5cR63) that the field may be missing in the message.
